File: lib_common/src/main/java/com/wujia/lib_common/data/network/exception/ExceptionEngine.java
package com.wujia.lib_common.data.network.exception;


import android.net.ParseException;

import com.google.gson.JsonParseException;

import org.json.JSONException;
import org.json.JSONObject;

import java.net.ConnectException;
import java.net.SocketTimeoutException;

import retrofit2.HttpException;


/**
 * Created by xmren on 2017/9/29.
 */

public class ExceptionEngine {

    //对应HTTP的状态码
    private static final int UNAUTHORIZED = 401;
    private static final int FORBIDDEN = 403;
    private static final int NOT_FOUND = 404;
    private static final int REQUEST_TIMEOUT = 408;
    private static final int INTERNAL_SERVER_ERROR = 500;
    private static final int BAD_GATEWAY = 502;
    private static final int SERVICE_UNAVAILABLE = 503;
    private static final int GATEWAY_TIMEOUT = 504;

    public static ApiException handleException(Throwable e) {


        ApiException ex;
        if (e instanceof HttpException) {             //HTTP错误
            HttpException httpException = (HttpException) e;
            switch (httpException.code()) {
                case UNAUTHORIZED:
                case FORBIDDEN:
                case NOT_FOUND:
                case REQUEST_TIMEOUT:
                case GATEWAY_TIMEOUT:
                case INTERNAL_SERVER_ERROR:
                case BAD_GATEWAY:
                case SERVICE_UNAVAILABLE:
                default:
                    ex = new ApiException(e, ERROR.HTTP_ERROR, "系统开小差，请稍后再试");
                    break;
            }
            return ex;
        } else if (e instanceof ApiJsonFormateException) {
            ApiJsonFormateException apiJsonFormateException = (ApiJsonFormateException) e;
            String originData = apiJsonFormateException.originData;
            try {
                JSONObject jsonObject = new JSONObject(originData);
                String code = jsonObject.optString("code");
                String msg = jsonObject.optString("msg");
                ex = new ApiException(e, code, msg);
            } catch (JSONException e1) {
                e1.printStackTrace();
                ex = new ApiException(e, ERROR.PARSE_ERROR, "数据解析异常");
            }
            return ex;
        } else if (e instanceof JsonParseException
                || e instanceof JSONException
                || e instanceof ParseException) {
            ex = new ApiException(e, ERROR.PARSE_ERROR, "数据解析异常");
            return ex;
        } else if (e instanceof ConnectException) {
            ex = new ApiException(e, ERROR.NETWORK_CONNECT_ERROR, "网络未连接，请检查网络");
            return ex;
        } else if (e instanceof SocketTimeoutException) {//超时
            ex = new ApiException(e, ERROR.NETWORK_TIMEOUT, "网络连接超时，请检查网络");
            return ex;
        } else if (e instanceof TokenException) {
            ex = new ApiException(e, ERROR.HTTP_ERROR, e.getMessage());
            return ex;
        } else {
            ex = new ApiException(e, ERROR.UNKNOWN, "未知错误");//未知错误
            return ex;
        }
    }
}

File: lib_common/src/main/java/com/wujia/lib_common/data/network/exception/ExceptionEngine.kt
package com.wujia.lib_common.data.network.exception


import android.net.ParseException
import com.google.gson.Gson
import com.google.gson.JsonParseException
import com.wujia.lib_common.base.RootResponse
import org.json.JSONException
import retrofit2.HttpException
import java.net.ConnectException
import java.net.SocketTimeoutException


/**
 * Created by xmren on 2017/9/29.
 */

object ExceptionEngine {

    //对应HTTP的状态码
    private val UNAUTHORIZED = 401
    private val FORBIDDEN = 403
    private val NOT_FOUND = 404
    private val REQUEST_TIMEOUT = 408
    private val INTERNAL_SERVER_ERROR = 500
    private val BAD_GATEWAY = 502
    private val SERVICE_UNAVAILABLE = 503
    private val GATEWAY_TIMEOUT = 504

    fun handleException(e: Throwable): ApiException {


        var ex: ApiException
        if (e is HttpException) {             //HTTP错误
            var msg: String? = "系统开小差，请稍后再试"
            try {
                val body = e.response().errorBody()!!.string()
                val response = Gson().fromJson(body, RootResponse::class.java)
                msg = response.message
            } catch (e1: Exception) {
                e1.printStackTrace()
            }

            when (e.code()) {
                UNAUTHORIZED, FORBIDDEN, NOT_FOUND, REQUEST_TIMEOUT, GATEWAY_TIMEOUT, INTERNAL_SERVER_ERROR, BAD_GATEWAY, SERVICE_UNAVAILABLE -> ex = ApiException(e, ERROR.HTTP_ERROR, msg)
                else -> ex = ApiException(e, ERROR.HTTP_ERROR, msg)
            }
            return ex
        } else if (e is ApiJsonFormateException) {
//            val originData = e.originData
//            try {
//                val jsonObject = JSONObject(originData)
//                val code = jsonObject.optString("code")
//                val msg = jsonObject.optString("msg")
//                ex = ApiException(e, code, msg)
//            } catch (e1: JSONException) {
//                e1.printStackTrace()
//            }
            e.printStackTrace()
            ex = ApiException(e, ERROR.PARSE_ERROR, "数据解析异常")

            return ex
        } else if (e is JsonParseException
                || e is JSONException
                || e is ParseException) {
            ex = ApiException(e, ERROR.PARSE_ERROR, "数据解析异常")
            return ex
        } else if (e is ConnectException) {
            ex = ApiException(e, ERROR.NETWORK_CONNECT_ERROR, "网络未连接，请检查网络")
            return ex
        } else if (e is SocketTimeoutException) {//超时
            ex = ApiException(e, ERROR.NETWORK_TIMEOUT, "网络连接超时，请检查网络")
            return ex
        } else if (e is TokenException) {
            ex = ApiException(e, ERROR.HTTP_ERROR, e.message)
            return ex
        } else {
            ex = ApiException(e, ERROR.UNKNOWN, "未知错误")//未知错误
            return ex
        }
    }
}

